tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+                constexpr size_t MaxDisallowedCharacters = 10;
+                if (disallowed_characters <= MaxDisallowedCharacters) {
+                    fprintf(stderr, "Validation failed: %s contains disallowed character 0x%02X.\n", filepath.c_str(),
+                        static_cast<unsigned int>(ch));
+                }
+            }
+
+            if ((prev == ' ' || prev == '\t') && (ch == CR || ch == LF)) {
+                ++trailing_whitespace_lines;
+            }
+
+            (void) exchange(previous3, exchange(previous2, exchange(prev, ch)));
+        }
+    }
+
+    if (prev == CR) { // file ends with CR
+        has_cr = true;
+    }
+
+    if (has_cr) {
+        fprintf(stderr, "Validation failed: %s contains CR line endings (possibly damaged CRLF).\n", filepath.c_str());
+    } else if (has_lf && has_crlf) {
+        fprintf(stderr, "Validation failed: %s contains mixed line endings (both LF and CRLF).\n", filepath.c_str());
+    } else if (has_lf) {
+        fprintf(stderr, "Validation failed: %s contains LF line endings.", filepath.c_str());
+
+        if (prev != LF) {
+            fprintf(stderr, " Also, it doesn't end with a newline.\n");
+        } else if (previous2 == LF) {
+            fprintf(stderr, " Also, it ends with multiple newlines.\n");
+        } else {
+            fprintf(stderr, "\n");
+        }
+    } else if (has_crlf) {
+        if (previous2 != CR || prev != LF) {
+            fprintf(stderr, "Validation failed: %s doesn't end with a newline.\n", filepath.c_str());
+        } else if (previous3 == LF) {
+            fprintf(stderr, "Validation failed: %s ends with multiple newlines.\n", filepath.c_str());
+        }
+    } else {
+        fprintf(stderr, "Validation failed: %s doesn't contain any newlines.\n", filepath.c_str());
+    }
+
+    if (has_utf8_bom) {
+        fprintf(stderr, "Validation failed: %s contains UTF-8 BOM characters.\n", filepath.c_str());
+    }
+
+    if (tab_characters != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu tab characters.\n", filepath.c_str(), tab_characters);
+    }
+
+    if (trailing_whitespace_lines != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu lines with trailing whitespace.\n", filepath.c_str(),
+            trailing_whitespace_lines);
+    }
+}
+
+int main() {
+    static constexpr array skipped_directories{
+        ".git"sv,
+        ".vs"sv,
+        ".vscode"sv,
+        "out"sv,
+        "vcpkg"sv,
+    };
+
+    static constexpr array skipped_extensions{
+        ".dll"sv,
+        ".exe"sv,
+        ".obj"sv,
+    };
+
+    // TRANSITION, P0202R3, use constexpr is_sorted()
+    assert(is_sorted(skipped_directories.begin(), skipped_directories.end()));
+    assert(is_sorted(skipped_extensions.begin(), skipped_extensions.end()));
+
+    vector<unsigned char> buffer; // reused for performance
+
+    try {
+        for (filesystem::recursive_directory_iterator rdi{"."}, last; rdi != last; ++rdi) {

```

Comment:
I think absolute paths might produce nicer output? (that is, filesystem::current_path() instead of ".")

This is locale sensitive so you probably want `u8path`.

Reply:
Regarding absolute vs. relative, I believe that the exact opposite is true - relative is nice, absolute is horrible. With relative paths, we get output like this (see #230):

```
Validation failed: .\CMakeSettings.json contains UTF-8 BOM characters.
Validation failed: .\docs\ISSUE_TEMPLATE.md contains LF line endings.
Validation failed: .\tools\CMakeSettings.json contains UTF-8 BOM characters.
Validation failed: .\vcpkg_windows.txt contains LF line endings.
```

This is relative to the source root, which seems ideal.

With absolute paths, irrelevant details from the Azure Pipelines directory structure will be printed.

Reply:
That's fair!